### PR TITLE
NEW Wire up symfony/validator

### PR DESCRIPTION
**Pull request by GuySartorelli**
## Description

Wires up `symfony/validator` so we can use it for robust and potentially complex validation logic.

Usage is:
```php
use SilverStripe\Core\Validation\ConstraintValidator;

/**
 * @var \Symfony\Component\Validator\Constraint $constraint
 * @var \SilverStripe\ORM\ValidationResult $result
 */
$result = ConstraintValidator::validate($valueToValidate, $constraint);
```

e.g. for testing if a URL is valid:
```php
use SilverStripe\Core\Validation\ConstraintValidator;
use Symfony\Component\Validator\Constraints\Url;

$isValid = ConstraintValidator::validate($url, new Url())->isValid();
```

Most of the constraints in https://symfony.com/doc/current/reference/constraints.html are explicitly supported, except for the following:

Any which require `symfony/expression-language`:
- `Symfony\Component\Validator\Constraints\Expression`
- `Symfony\Component\Validator\Constraints\ExpressionSyntax`
- `Symfony\Component\Validator\Constraints\When`

Any which require `symfony/http-client`:
- `Symfony\Component\Validator\Constraints\NotCompromisedPassword`

Any which require `symfony/intl`:
- `Symfony\Component\Validator\Constraints\Bic`
- `Symfony\Component\Validator\Constraints\Country`
- `Symfony\Component\Validator\Constraints\Currency`
- `Symfony\Component\Validator\Constraints\Language`
- `Symfony\Component\Validator\Constraints\Locale`

Any which require `symfony/doctrine` or anything which assumes we're using symfony/validate in the context of setting constraints on metadata objects:
- `Symfony\Component\Validator\Constraints\Cascade`
- `Symfony\Component\Validator\Constraints\Traverse`
- `Symfony\Component\Validator\Constraints\Valid`

With exception of those last three (which I've mentioned in the docs are explicitly not supported, because it relies on extra wiring up of things that we just aren't gonna do), I'd expect the others to all work - I just haven't included tests for them because we don't have those dependencies.

## Issues

- https://github.com/silverstripe/silverstripe-linkfield/issues/17

## Pull request checklist

- [x] The target branch is correct
    - See [picking the right version](https://docs.silverstripe.org/en/contributing/code/#picking-the-right-version)
- [x] All commits are relevant to the purpose of the PR (e.g. no debug statements, unrelated refactoring, or arbitrary linting)
    - Small amounts of additional linting are usually okay, but if it makes it hard to concentrate on the relevant changes, ask for the unrelated changes to be reverted, and submitted as a separate PR.
- [x] The commit messages follow our [commit message guidelines](https://docs.silverstripe.org/en/contributing/code/#commit-messages)
- [x] The PR follows our [contribution guidelines](https://docs.silverstripe.org/en/contributing/code/)
- [x] Code changes follow our [coding conventions](https://docs.silverstripe.org/en/contributing/coding_conventions/)
- [x] This change is covered with tests (or tests aren't necessary for this change)
- [x] Any relevant User Help/Developer documentation is updated; for impactful changes, information is added to the changelog for the intended release
- [x] CI is green

### CODE REVIEW

File: src/Core/Validation/ConstraintValidator.php [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+/**
+ * Helper class to abstract away wiring up symfony/validator and getting ValidationResult from validating
+ * symfony validator constraints.
+ */
+class ConstraintValidator
+{
+    /**
+     * Validate a value by a constraint
+     *
+     * @param Constraint[]|Constraint $constraints a constraint or array of constraints to validate against
```
Comment (emteknetnz) [inline src/Core/Validation/ConstraintValidator.php]:
```suggestion
     * @param Constraint|Constraint[] $constraints a constraint or array of constraints to validate against
```

**Review by emteknetnz — CHANGES_REQUESTED** [review]
Add a unit test(s) for passing in an array of contraints

Add a unit test for passing in optional field param

File: src/Core/Validation/ConstraintValidator.php (hunk 1 above)
Comment (GuySartorelli) [reply src/Core/Validation/ConstraintValidator.php]: Done

**Comment by GuySartorelli** [comment]
> Add a unit test(s) for passing in an array of contraints

Done

> Add a unit test for passing in optional field param

Done

Also added an exception (and test for it) when passing in an empty array.

**Review by emteknetnz — APPROVED** [review]
